### PR TITLE
Update workflow state without using painless script 

### LINKED ISSUE #779: [REFACTOR] Improve Workflow State Resource updates
### Is your feature request related to a problem?

Currently the Workflow State resource document is only additive, using a Painless script during provisioning to append to a list.  This prevented iteratively updating it during deprovisioning, leading to open questions like #691.

Following #778 we will have the ability to update any individual element of the Workflow State which will bring these benefits:
 - Allow for incremental addition to the resources created without using Painless, improving portability
 - Allow for incremental removal from the resources created during deprovisioning, allowing the full deprovisioning to depend on this update. (See #790) 
 - Allow for the easy removal of the "error" field on subsequent successful provisioning

### What solution would you like?

Replace the painless script with appropriate code from #778.  Use `ifSeqNo()` and `ifPrimaryTerm()` on the update requests to enforce concurrency, and retry if it fails.

### PR DESCRIPTION
**Pull request by dbwiddis**
### Description

Removes methods using painless script for workflow state updates.

### Related Issues

Fixes #779 (except for the part that will be covered by #780)

### Check List
- [x] New functionality includes testing.
- [x] Commits are signed per the DCO using `--signoff`.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/flow-framework/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/flowframework/indices/FlowFrameworkIndicesHandler.java [hunk 1]
```diff
@@ -716,26 +679,93 @@ public void addResourceToStateIndex(
...
+        if (!doesIndexExist(WORKFLOW_STATE_INDEX)) {
+            String errorMessage = "Failed to update state for " + workflowId + " due to missing " + WORKFLOW_STATE_INDEX + " index";
+            logger.error(errorMessage);
+            listener.onFailure(new FlowFrameworkException(errorMessage, RestStatus.NOT_FOUND));
+        } else {
+            try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
+                getAndUpdateResourceInStateDocumentWithRetries(
+                    workflowId,
+                    newResource,
+                    3,
```
Comment (amitgalitz) [inline src/main/java/org/opensearch/flowframework/indices/FlowFrameworkIndicesHandler.java:694]: Is 3 retries a standard? we use 5 in other locations

**Review by joshpalis — APPROVED** [review]

File: src/main/java/org/opensearch/flowframework/indices/FlowFrameworkIndicesHandler.java [hunk 2]
```diff
@@ -716,26 +679,93 @@ public void addResourceToStateIndex(
...
+    private void handleStateUpdateException(
+        String workflowId,
+        ResourceCreated newResource,
+        int retries,
+        ActionListener<WorkflowData> listener,
+        Exception e
+    ) {
+        if (e instanceof VersionConflictEngineException && retries > 0) {
+            // Retry if we haven't exhausted retries
+            getAndUpdateResourceInStateDocumentWithRetries(workflowId, newResource, retries - 1, listener);
```
Comment (owaiskazi19) [inline src/main/java/org/opensearch/flowframework/indices/FlowFrameworkIndicesHandler.java:757]: Why this recursive calling here when we can pass retryonconflict?

File: src/test/java/org/opensearch/flowframework/indices/FlowFrameworkIndicesHandlerTests.java [hunk 3]
```diff
@@ -537,6 +606,147 @@ public void testAddResourceToStateIndex() throws IOException {
...
+        }).doAnswer(invocation -> {
+            ActionListener<UpdateResponse> responseListener = invocation.getArgument(1);
+            responseListener.onFailure(conflictException);
+            return null;
+        }).doAnswer(invocation -> {
+            // we'll never get here
+            ActionListener<UpdateResponse> responseListener = invocation.getArgument(1);
+            responseListener.onResponse(updateResponse);
+            return null;
+        }).when(client).update(any(UpdateRequest.class), any());
```
Comment (owaiskazi19) [inline src/test/java/org/opensearch/flowframework/indices/FlowFrameworkIndicesHandlerTests.java:735]: This is great. We should move this to a common test class for other tests to use retry listeners

**Review by owaiskazi19 — COMMENTED** [review]
LGTM with minor comments

File: src/main/java/org/opensearch/flowframework/indices/FlowFrameworkIndicesHandler.java [hunk 4]
```diff
@@ -608,99 +610,60 @@ public void updateFlowFrameworkSystemIndexDoc(
...
         } else {
             try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
-                DeleteRequest deleteRequest = new DeleteRequest(WORKFLOW_STATE_INDEX, documentId);
-                deleteRequest.setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
-                client.delete(deleteRequest, ActionListener.runBefore(listener, context::restore));
+                UpdateRequest updateRequest = new UpdateRequest(WORKFLOW_STATE_INDEX, documentId);
+                Map<String, Object> updatedContent = new HashMap<>(updatedFields);
+                updateRequest.doc(updatedContent);
+                updateRequest.setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
+                updateRequest.retryOnConflict(5);
```
[the remark this answers is not in the document]
Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/indices/FlowFrameworkIndicesHandler.java:633]: I don't know why they are different and I don't think conflicts are going to be common.  These are updates to a single workflow's state index, and the only time there would be a conflict is if two (or more) resources are simultaneously trying to update the resource list.  I'd have to think this is an extremely rare situation, if at all.  Sure, I can make it a constant.

File: src/main/java/org/opensearch/flowframework/indices/FlowFrameworkIndicesHandler.java (hunk 2 above)
Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/indices/FlowFrameworkIndicesHandler.java:757]: RetryOnConflict only applies when you don't care what the existing value is but you're updating a field.  In our case we need to fetch the existing resource list, add to it, and then put that resource list back.  

If the document changed in between us getting and updating, retryOnConflict would miss it.  We use the `seqNo` and `primaryTerm` from the document we "got" to make sure that's the same document we're putting back in; if the doc has been updated, retrying won't do us any good, we need to retry the whole get-and-put which is what this does.

Also a minor note, while it looks recursive, it's just a chained callback not unlike all our other transport calls, so it won't fill up the stack.

File: src/test/java/org/opensearch/flowframework/indices/FlowFrameworkIndicesHandlerTests.java (hunk 3 above)
Comment (dbwiddis) [reply src/test/java/org/opensearch/flowframework/indices/FlowFrameworkIndicesHandlerTests.java:735]: Not sure what you mean here?

Comment (owaiskazi19) [reply src/test/java/org/opensearch/flowframework/indices/FlowFrameworkIndicesHandlerTests.java:735]: I was thinking if we can have a single method which can be used across tests to test retry listeners but I don't think so there would be more case. So it should be fine

File: src/main/java/org/opensearch/flowframework/indices/FlowFrameworkIndicesHandler.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/main/java/org/opensearch/flowframework/indices/FlowFrameworkIndicesHandler.java]: Should we update the `3` also with the constant now?

**Review by owaiskazi19 — APPROVED** [review]

**Review by amitgalitz — APPROVED** [review]

**Pull request merged**